search-books/src/main/java/nix/education/service/BookDataService.java

```diff
@@ -0,0 +1,43 @@
+package nix.education.service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import nix.education.entity.Book;
+import nix.education.entity.Tag;
+import nix.education.repository.BookRepository;
+import nix.education.repository.TagRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class BookDataService {
+
+    private TagRepository tagRepository;
+
+    public void setTags(Book book) {
+        List<Tag> bookTags = book.getTags();
+        List<String> tagTitles = getTagTitles(book);
+        List<Tag> existingTags = tagRepository.findTagsByTagIn(tagTitles);
+        existingTags.forEach(
+            existedTag -> bookTags.removeIf(tag -> tag.getTag().equals(existedTag.getTag())));
+
+        List<Tag> resolvedTags = new ArrayList<>();
+        resolvedTags.addAll(existingTags);
+        resolvedTags.addAll(bookTags);
+
+        book.setTags(resolvedTags);
+        book.setTagTitles(tagTitles);
+    }
+
+    public List<String> getTagTitles(Book book) {
+        return book.getTags().stream().map(Tag::getTag).collect(Collectors.toList());
+    }
+
+    public List<Book> getBooksByIds(List<Long> ids, BookRepository bookRepository) {
+        List<Book> books = new ArrayList<>();
+        ids.forEach(id -> books.add(bookRepository.findById(id).orElseThrow()));
```

Comment:
Same problem as we had for tags. One call to DB per id - not good. Can we fetch all the required docs by a single query?

Reply:
Yes. I agree. Refactored.